File: products/dcmaar/modules/desktop/src/utils/CircuitBreaker.ts
/**
 * Circuit Breaker pattern for resilience
 * Prevents cascading failures by opening circuit after threshold failures
 */

export enum CircuitState {
  CLOSED = 'CLOSED', // Normal operation
  OPEN = 'OPEN', // Failing, reject requests
  HALF_OPEN = 'HALF_OPEN', // Testing if service recovered
}

export interface CircuitBreakerConfig {
  failureThreshold: number; // Number of failures before opening
  resetTimeout: number; // Time in ms before attempting half-open
  halfOpenThreshold: number; // Number of successes in half-open to close
  monitoringWindow: number; // Time window for failure counting (ms)
}

export const DEFAULT_CIRCUIT_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  resetTimeout: 30_000, // 30 seconds
  halfOpenThreshold: 3,
  monitoringWindow: 60_000, // 1 minute
};

export interface CircuitBreakerStats {
  state: CircuitState;
  failures: number;
  successes: number;
  lastFailureTime?: Date;
  lastSuccessTime?: Date;
  nextAttemptTime?: Date;
}

/**
 * Circuit Breaker implementation
 */
export class CircuitBreaker {
  private state: CircuitState = CircuitState.CLOSED;
  private failures: number = 0;
  private successes: number = 0;
  private lastFailureTime?: Date;
  private lastSuccessTime?: Date;
  private nextAttemptTime?: Date;
  private failureTimestamps: number[] = [];

  constructor(private config: CircuitBreakerConfig = DEFAULT_CIRCUIT_CONFIG) {}

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (!this.canAttempt()) {
      throw new CircuitBreakerError(
        `Circuit breaker is ${this.state}`,
        this.state
      );
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  /**
   * Check if a request can be attempted
   */
  private canAttempt(): boolean {
    switch (this.state) {
      case CircuitState.CLOSED:
        return true;

      case CircuitState.OPEN:
        // Check if reset timeout has elapsed
        if (this.nextAttemptTime && Date.now() >= this.nextAttemptTime.getTime()) {
          this.state = CircuitState.HALF_OPEN;
          this.successes = 0;
          return true;
        }
        return false;

      case CircuitState.HALF_OPEN:
        return true;

      default:
        return false;
    }
  }

  /**
   * Record a successful execution
   */
  private onSuccess(): void {
    this.lastSuccessTime = new Date();
    this.failures = 0;
    this.failureTimestamps = [];

    if (this.state === CircuitState.HALF_OPEN) {
      this.successes++;
      if (this.successes >= this.config.halfOpenThreshold) {
        this.state = CircuitState.CLOSED;
        this.successes = 0;
      }
    }
  }

  /**
   * Record a failed execution
   */
  private onFailure(): void {
    this.lastFailureTime = new Date();
    const now = Date.now();

    // Add failure timestamp
    this.failureTimestamps.push(now);

    // Remove old failures outside monitoring window
    this.failureTimestamps = this.failureTimestamps.filter(
      (timestamp) => now - timestamp < this.config.monitoringWindow
    );

    this.failures = this.failureTimestamps.length;

    if (this.state === CircuitState.HALF_OPEN) {
      // Immediately open if failure in half-open state
      this.openCircuit();
    } else if (this.failures >= this.config.failureThreshold) {
      this.openCircuit();
    }
  }

  /**
   * Open the circuit
   */
  private openCircuit(): void {
    this.state = CircuitState.OPEN;
    this.nextAttemptTime = new Date(Date.now() + this.config.resetTimeout);
    this.successes = 0;
  }

  /**
   * Manually reset the circuit breaker
   */
  reset(): void {
    this.state = CircuitState.CLOSED;
    this.failures = 0;
    this.successes = 0;
    this.failureTimestamps = [];
    this.lastFailureTime = undefined;
    this.lastSuccessTime = undefined;
    this.nextAttemptTime = undefined;
  }

  /**
   * Get current circuit breaker statistics
   */
  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failures: this.failures,
      successes: this.successes,
      lastFailureTime: this.lastFailureTime,
      lastSuccessTime: this.lastSuccessTime,
      nextAttemptTime: this.nextAttemptTime,
    };
  }

  /**
   * Get current state
   */
  getState(): CircuitState {
    return this.state;
  }

  /**
   * Check if circuit is open
   */
  isOpen(): boolean {
    return this.state === CircuitState.OPEN;
  }

  /**
   * Check if circuit is closed
   */
  isClosed(): boolean {
    return this.state === CircuitState.CLOSED;
  }
}

/**
 * Circuit Breaker specific error
 */
export class CircuitBreakerError extends Error {
  constructor(
    message: string,
    public readonly state: CircuitState
  ) {
    super(message);
    this.name = 'CircuitBreakerError';
  }
}

File: products/yappc/infrastructure/datacloud/src/main/java/com/ghatana/yappc/infrastructure/resilience/CircuitBreaker.java
package com.ghatana.yappc.infrastructure.resilience;

import io.activej.promise.Promise;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

/**
 * Circuit breaker pattern implementation for resilient service calls.
 *
 * <p><b>Purpose</b><br>
 * Prevents cascading failures by temporarily rejecting requests when a service
 * is experiencing high error rates. Automatically transitions between states
 * based on success/failure metrics.
 *
 * <p><b>States</b><br>
 * - CLOSED: Normal operation, requests pass through<br>
 * - OPEN: Failure threshold exceeded, requests fail fast<br>
 * - HALF_OPEN: Testing if service has recovered<br>
 *
 * <p><b>Usage</b><br>
 * <pre>{@code
 * CircuitBreaker breaker = CircuitBreaker.builder()
 *     .failureThreshold(5)
 *     .recoveryTimeout(Duration.ofSeconds(30))
 *     .halfOpenMaxCalls(3)
 *     .build();
 *
 * // Wrap AI service call
 * Promise<Result> result = breaker.execute(() -> aiService.complete(request));
 * }</pre>
 *
 * @doc.type class
 * @doc.purpose Circuit breaker for fault tolerance
 * @doc.layer infrastructure
 * @doc.pattern Resilience, Circuit Breaker
 */
public final class CircuitBreaker {

    private static final Logger LOG = LoggerFactory.getLogger(CircuitBreaker.class);

    private final int failureThreshold;
    private final Duration recoveryTimeout;
    private final int halfOpenMaxCalls;
    private final double successThreshold;

    private final AtomicReference<State> state = new AtomicReference<>(State.CLOSED);
    private final AtomicInteger failureCount = new AtomicInteger(0);
    private final AtomicInteger successCount = new AtomicInteger(0);
    private final AtomicInteger halfOpenCalls = new AtomicInteger(0);
    private final AtomicLong lastFailureTime = new AtomicLong(0);

    private CircuitBreaker(Builder builder) {
        this.failureThreshold = builder.failureThreshold;
        this.recoveryTimeout = builder.recoveryTimeout;
        this.halfOpenMaxCalls = builder.halfOpenMaxCalls;
        this.successThreshold = builder.successThreshold;
    }

    /**
     * Executes the given supplier with circuit breaker protection.
     *
     * @param supplier the operation to execute
     * @param <T> the result type
     * @return Promise of the operation result
     * @throws CircuitBreakerOpenException if circuit is OPEN
     */
    public <T> Promise<T> execute(Supplier<Promise<T>> supplier) {
        State currentState = state.get();

        switch (currentState) {
            case OPEN:
                if (shouldAttemptReset()) {
                    transitionTo(State.HALF_OPEN);
                    return executeHalfOpen(supplier);
                }
                LOG.debug("Circuit breaker OPEN - failing fast");
                return Promise.ofException(new CircuitBreakerOpenException(
                    "Service temporarily unavailable", getRemainingTimeout()));

            case HALF_OPEN:
                return executeHalfOpen(supplier);

            case CLOSED:
            default:
                return executeClosed(supplier);
        }
    }

    private <T> Promise<T> executeClosed(Supplier<Promise<T>> supplier) {
        try {
            return supplier.get()
                .whenResult(result -> {
                    // Success - reset failure count
                    failureCount.set(0);
                })
                .whenException(exception -> {
                    // Failure - increment count and check threshold
                    int failures = failureCount.incrementAndGet();
                    lastFailureTime.set(System.currentTimeMillis());
                    if (failures >= failureThreshold) {
                        LOG.warn("Circuit breaker threshold exceeded ({} failures), transitioning to OPEN", failures);
                        transitionTo(State.OPEN);
                    }
                });
        } catch (Exception e) {
            failureCount.incrementAndGet();
            lastFailureTime.set(System.currentTimeMillis());
            return Promise.ofException(e);
        }
    }

    private <T> Promise<T> executeHalfOpen(Supplier<Promise<T>> supplier) {
        int currentHalfOpenCalls = halfOpenCalls.incrementAndGet();
        if (currentHalfOpenCalls > halfOpenMaxCalls) {
            return Promise.ofException(new CircuitBreakerOpenException(
                "Half-open call limit exceeded", getRemainingTimeout()));
        }

        try {
            return supplier.get()
                .whenResult(result -> {
                    // Success in half-open - increment success count
                    int successes = successCount.incrementAndGet();
                    halfOpenCalls.decrementAndGet();
                    if (successes >= halfOpenMaxCalls * successThreshold) {
                        LOG.info("Circuit breaker recovered, transitioning to CLOSED");
                        transitionTo(State.CLOSED);
                    }
                })
                .whenException(exception -> {
                    // Failure in half-open - go back to OPEN
                    halfOpenCalls.decrementAndGet();
                    LOG.warn("Circuit breaker recovery failed, transitioning back to OPEN");
                    transitionTo(State.OPEN);
                });
        } catch (Exception e) {
            halfOpenCalls.decrementAndGet();
            transitionTo(State.OPEN);
            return Promise.ofException(e);
        }
    }

    private boolean shouldAttemptReset() {
        long lastFailure = lastFailureTime.get();
        return System.currentTimeMillis() - lastFailure >= recoveryTimeout.toMillis();
    }

    private Duration getRemainingTimeout() {
        long lastFailure = lastFailureTime.get();
        long elapsed = System.currentTimeMillis() - lastFailure;
        long remaining = Math.max(0, recoveryTimeout.toMillis() - elapsed);
        return Duration.ofMillis(remaining);
    }

    private void transitionTo(State newState) {
        State oldState = state.getAndSet(newState);
        if (oldState != newState) {
            LOG.info("Circuit breaker state transition: {} -> {}", oldState, newState);
            // Reset counters on state change
            failureCount.set(0);
            successCount.set(0);
            halfOpenCalls.set(0);
        }
    }

    /**
     * Returns current circuit breaker state.
     *
     * @return current state
     */
    public State getState() {
        return state.get();
    }

    /**
     * Manually reset circuit breaker to CLOSED state.
     */
    public void reset() {
        transitionTo(State.CLOSED);
        lastFailureTime.set(0);
        LOG.info("Circuit breaker manually reset to CLOSED");
    }

    /**
     * Circuit breaker states.
     */
    public enum State {
        CLOSED,     // Normal operation
        OPEN,       // Failing fast
        HALF_OPEN   // Testing recovery
    }

    /**
     * Creates a new builder for circuit breaker configuration.
     *
     * @return builder instance
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Builder for CircuitBreaker.
     */
    public static final class Builder {
        private int failureThreshold = 5;
        private Duration recoveryTimeout = Duration.ofSeconds(30);
        private int halfOpenMaxCalls = 3;
        private double successThreshold = 0.5;

        private Builder() {}

        public Builder failureThreshold(int threshold) {
            this.failureThreshold = threshold;
            return this;
        }

        public Builder recoveryTimeout(Duration timeout) {
            this.recoveryTimeout = timeout;
            return this;
        }

        public Builder halfOpenMaxCalls(int maxCalls) {
            this.halfOpenMaxCalls = maxCalls;
            return this;
        }

        public Builder successThreshold(double threshold) {
            this.successThreshold = threshold;
            return this;
        }

        public CircuitBreaker build() {
            return new CircuitBreaker(this);
        }
    }
}
